zoekt: use b-tree for filenames

stefanhengl opened the pull request:
With this change we use the b-tree for filnames just like we do for content. Based on heap profiles from production instances, we anticipate a reduction of the heap by 15-20%. The change is behind the feature flag "ZOEKT_ENABLE_BTREE_NAME". 

For now, we use a makeshift wrapper just for the feature-flag (see comments in code). This is all temporary because I expect us to use the concrete type `btreeIndex` in indexdata as soon as we have validated the new code.

Test plan:
- With the exception of trivial tests for sizeBytes, all tests pass with and without the feature flag.  
- pprof

started webserver on a corpus of ~630 repos with and without ZOEKT_ENABLE_BTREE_NAME="1" 

```
ZOEKT_ENABLE_BTREE_NAME="1" ZOEKT_ENABLE_BTREE="1"  go run . --pprof --index ~/corpus/index
```

and compared the heap profiles, especially w.r.t. zoekt.(*indexdata).readFileNameNgrams

```
go tool pprof  -http=localhost:6073 http://localhost:6070/debug/pprof/heap
```

before:
<img width="1665" alt="Screenshot 2023-02-08 at 17 51 35" src="https://user-images.githubusercontent.com/26413131/217597679-4a0ba0fe-0f71-4158-9b1e-d72a32e76cdb.png">

after:
<img width="1663" alt="Screenshot 2023-02-08 at 17 51 26" src="https://user-images.githubusercontent.com/26413131/217597995-2576a672-f0de-4050-aa03-10c99eda65c7.png">

keegancsmith commented on btree.go:
```diff
@@ -308,6 +308,10 @@ func (b btreeIndex) SizeBytes() (sz int) {
 func (b btreeIndex) Get(ng ngram) (ss simpleSection) {
+	if b.bt == nil {
+		return simpleSection{}
+	}
```
is this fixing a bug in main?

Edit: I think it is from ngramFrequency were for ngram we handle it being nil but in the new code you don't do it there?

stefanhengl replied:
This is just in anticipation of us using the concrete type IE btreeIndex in indexdata, in which case a well behaved 0-value is a nice-to-have for tests. In production we always contruct the btreeIndex with newBtreeIndex which sets the btree correctly.

keegancsmith commented on hititer.go:
```diff
@@ -120,23 +120,25 @@ func (d *indexData) trigramHitIterator(ng ngram, caseSensitive, fileName bool) (
+	if fileName {
+		buildIters(d.fileNameNgrams)
+	} else {
+		buildIters(d.ngrams)
```
minor: I have no idea if this is more efficient, but you can avoid the closure if you do something like

```
var ngrams ngramIndex
if fileName {
  ngrams = d.fileNameNgrams
} else {
  ngrams = d.ngrams
}
...
```

keegancsmith commented on read.go:
```diff
@@ -321,7 +320,7 @@ func (r *reader) readIndexData(toc *indexTOC) (*indexData, error) {
 	d.fileNameIndex = toc.fileNames.relativeIndex()
 
-	d.fileNameNgrams, err = d.readFileNameNgrams(toc)
+	d.fileNameNgrams, err = d.newBtreeIndex(toc.nameNgramText, toc.namePostings)
```
I guess before this leaves draft this will be upgraded to a feature flag?

stefanhengl committed:
use btree for fileNameNgrams

stefanhengl committed:
check for b.bt == nil

This is mainly in anticipation of using btreeIndex as the default and to
make sure that the 0-value of a btreeIndex doesn't panic in tests. In
production we always use the constructor, indexdata.newBtreeIndex, which
always sets b.bt.

stefanhengl committed:
remove redundant conversion

keegancsmith reviewed (approved):
so much excite
